src/assets/TokensController.ts

```diff
@@ -216,32 +238,48 @@ export class TokensController extends BaseController<
     const releaseLock = await this.mutex.acquire();
     try {
       address = toChecksumHexAddress(address);
-      const { tokens, ignoredTokens } = this.state;
+      const { tokens, ignoredTokens, detectedTokens } = this.state;
+      let newTokens: Token[] = [...tokens];
       const isERC721 = await this._detectIsERC721(address);
-      const newEntry: Token = { address, symbol, decimals, image, isERC721 };
-      const previousEntry = tokens.find(
+      const tokenMetadata = await this.fetchTokenMetadata(address);
```

Comment:
If the current network changes _after_ `addToken` is called but before this returns, the token will get added to the wrong network's token list. This is a pre-existing problem because of the `await` on the `_detectIsERC721` call, but it's still made worse here by this second `await`.

This could be avoided by caching the current `chainId` when this function starts, and throwing an error if the `chainId` has changed after this call.

Reply:
Updated code with this change. Also included a test here - https://github.com/MetaMask/controllers/pull/808/commits/638eead3e0c0515aa1e14f05d59fac87f5ba8af9